### ISSUE TITLE
add -fstack-clash-protection gcc flag

### ISSUE THREAD

**Issue by frank-dittrich** [magnumripper/JohnTheRipper#3599]
Adding

**Comment by frank-dittrich** [comment]
Adding `-fstack-clash-protection` gcc flag might be a good idea, see
https://www.qualys.com/2019/01/09/system-down/system-down.txt
``` 
To the best of our knowledge, all systemd-based Linux distributions are
vulnerable, but SUSE Linux Enterprise 15, openSUSE Leap 15.0, and Fedora
28 and 29 are not exploitable because their user space is compiled with
GCC's -fstack-clash-protection.
```

The change is as simple as this:
```
diff --git a/src/configure.ac b/src/configure.ac
index b92f07434..7beb41403 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -269,6 +269,8 @@ fi
    JTR_FLAG_CHECK([-std=gnu89], 1)
    dnl We like reproducible builds
    JTR_FLAG_CHECK([-Wdate-time], 1)
+   dnl try to protect against stack-clash attacks
+   JTR_FLAG_CHECK([-fstack-clash-protection], 1)
    JTR_LIST_ADD(CFLAGS_EXTRA, ["$CFLAGS_EX"])
 ],
    AC_MSG_NOTICE([Unable to validate $CC command line arguments. Using standard values.])
```
(plus the resulting changes to src/configure)

We need to check for a possible performance impact.

**Comment by solardiz** [comment]
I think we should have a policy to avoid `alloca`, VLAs, and large stack arrays. With that policy, do we really want to also have runtime checks by default? I guess a question may be how large arrays are still allowed on stack - if we might allow a few arrays of a few KB each in one stack frame, then that has potential to jump over a guard page.

**Comment by solardiz** [comment]
I'm OK with having this in a debug build. I think we probably shouldn't have it in a default production build because JtR is special in that it can't tolerate almost any performance impact whereas the likelihood of a stack clash actually being the attack on JtR in the wild and it succeeding is negligible. If someone were to attack JtR with malicious input in the wild, which is unlikely but maybe not negligibly so, they'd likely find a more reliable attack vector using some format-specific bug or even a feature (if the malicious input is, say, a configuration file).

We can have a non-default configure option enabling a hardened build, which would be documented as having slight performance impact. We'd also enable `-fstack-protector` there, which would likely have greater performance impact (IIRC, I measured around 1% average with `relbench` a few years ago) yet also greater relevance.

**Comment by solardiz** [comment]
Found my old `-fstack-protector` benchmarks with gcc 4.6.x on Core 2, x86_64 builds only:

https://www.openwall.com/lists/owl-dev/2011/11/07/1

There was no measurable performance impact on JtR jumbo overall from `-fstack-protector`, but there was a 1.5% impact from `-fstack-protector-all`. Apparently, later I determined that `-fstack-protector --param=ssp-buffer-size=2` is also affordable (as in: almost no performance impact) as a default for Owl, but we never actually made it the default. Maybe that's what we can use for moderately hardened builds of JtR (and heavily hardened ones could use `-fstack-protector-all` despite of its 1.5% impact).

Of course, these things are gcc version and arch specific, and performance impact may vary greatly between JtR formats, etc.

**Comment by solardiz** [comment]
Another aspect to consider is what message we will be sending and whether it's what we mean. Will this appear to say that we mean JtR for processing of untrusted input and that's why we bother to harden it? Or will it mean that JtR's code is so poor that only hardening can help? Or what I think we really mean - that JtR is not meant for processing of untrusted input, but we acknowledge that it might be used like that anyway and we try to improve its code quality (even if to a limited extent) and we also provide some optional defense-in-depth for such risky uses that we think occur at users' own risk?

**Comment by jfoug** [comment]
> I'm OK with having this in a debug build.

The problem with this deploy logic being done due to 'lack of testing for performance regression'  is that now you have placed the code to only be built into code which CAN NOT be tested for performance ;)

We already have a ``` SecureMode = N ``` mode in john.conf.  Why do we not simply implement this into a ``` --enable-secure ``` configure build, which adds this ``` -fstack-clash-protection ``` AND sets the configure (through a ./run/john-local.conf or something) to be ``` SecureMode = Y ```

Or would these be 2 separate 'modes' which should not be handled as a pair?  The SecureMode=x is a feature done while I was away, so I do not know all of the background on it.  It may deal more with 'security' of the .pot file and output trying to avoid leaking cleartext passwords and have little/nothing to do with the security of the process itself.

**Comment by jfoug** [comment]
I think the ``` --enable-hardening ``` configure.ac option is much preferable to simply adding this new flag to a ASAN/Debug build.   Yes, the end user probably already knows that hardening can make a small reduction in performance.  That warning would be output in a ./configure run, by placing the 'Hardened code' into the section along with ASAN/UBSAN/DEBUG etc which already list to the end user that the program may take a performance hit.

If done in that manner, then we could easily release this flag without having to 'wait' for regression testing, which will NEVER be complete enough anyway.  The end user now knows that this feature IS available, and will build that way, if they choose to.

@frank-dittrich (or whoever produces this), we should also add this to the output of a ```--list=build-options``` so that it is easy to determine that the john image was built using this new feature.  So there will be an ```#define HAVE_HARDENING``` flag added by configure, and section(s) of code in listconf.c will need to be added to process this flag listing the hardened process.

**Comment by frank-dittrich** [comment]
@jfoug 
I agree with adding `--enable-hardening` and adding `-fstack-clash-protection` only for hardened builds.
Or, `--with-hardening=<level>`.
There will be different hardening options with individual performance impact.
The impact of `-fstack-clash-protection` should be smaller than that of  `-fstack-protector`, especially on recent linux systems where the OS provides the necessary infrastructure.

I am not sure yet when I'll have time to adjust my pull request https://github.com/magnumripper/JohnTheRipper/pull/3600.
HAVE_HARDENING should only be defined if the compiler supports any of the options we want to add with `--enable-hardening`.
If you want to report those hardening options, why not any CFLAGS_EXTRA settings the user specified?
